### ISSUE TITLE
`powd` overflow

### ISSUE THREAD

**Issue by maurolacy** [paupino/rust-decimal#431]
Hello, and first off, thanks for the library.

We're using `rust_decimal` to compute a function using fixed point math, involving `exp` and `powd`.

Now, what we are noticing is that `powd`'s range is not very large. By example, when raising a `Decimal` larger than 32_313_447 to a Decimal power smaller than one (0.68 in particular) using `powd`, we are hitting

`thread ... panicked at 'Pow overflowed', .../github.com-1ecc6299db9ec823/rust_decimal-1.16.0/src/maths.rs:283:21`

This doesn't make much sense, as a number raised to an exponent smaller than 1 would yield a _smaller_ number (127_755, in this particular case).

Are you aware of this? I can always take a look at the code, but, asking first in case this is a kwown issue / there are known workarounds. What do you suggest? 

Thanks,

**Comment by maurolacy** [comment]
Also, I've noticed that `sqrt()` doesn't suffer from this. Or at least, its range of application is order of magnitudes larger than `powd`. So, `powd` can for sure be made to behave, at least for exponents smaller than one.

**Comment by schungx** [comment]
> We cannot. We are forbidden to use floating point at all, because of non-determinism, i.e. different results being obtained
> in different implementations / architectures.

I understand that the IEEE 754 standard defines deterministic calculations for addition, subtraction, multiplication, division plus the square root, meaning that they should generate deterministic output when given the exact same inputs.

For exponential (and also power), you start running into the Table Maker's Dilemma and sooner or later you get into rounding situations.

It is probably possible to make an `exp` or `pow` where, for each round, it then normalizes the number into a fixed number of significant digits before continuing... which may make it accept a wider range of inputs at the risk of having rounding errors starts accumulating.

I'm not really an expert on this, but if you want `exp` or `pow` to work over a full range of 32-bit numbers, you may need more than the 96 bits that this crate is keeping.  In fact, you may need to use a arbitrary-length number library.

**Comment by schungx** [comment]
https://github.com/paupino/rust-decimal/blob/master/src/maths.rs#L316-323

This is where it calculates the `pow` by converting it into `exp`, and this may be the place where it overflows.

Since you know the exponent will be <1, I suggest you break down the steps and display intermediate values to find out which step it is overflowing.  Then you may simply add `normalize` calls?

**Comment by maurolacy** [comment]
OK, best solution for me is to dump the last term (27) of the power series when computing the exponential. That introduces an error of roughly one decimal digit, but extends the range of `powd` by about 40 times.

That said, I don't think this is the right thing to do. Will just introduce an upper cut-off / saturating value in my function, and that's it.

Thanks for your time and your work.

**Comment by schungx** [comment]
We can also use different algorithms for different ranges of inputs, picking the best one depending on the actual input values.

**Comment by dovahcrow** [comment]
We got a similar issue with the following code:
```rust
>> let v = Decimal::from_f64(13.815517970299580976037625513).unwrap();
>> v.exp()
```

We are also using Decimal in the blockchain.

**Comment by paupino** [comment]
@arrtchiu For division to overflow it would typically mean the quotient was being divided by a very small number which is a slightly different problem. Behind the scenes, division actually does what `powd` _should_ be doing - it expands the result to 192 bits and then tries to strip the underflow so that it can fit back into the 96 bit mantissa. In other words, it's unlikely to be the same problem - namely because it would be a true overflow if we couldn't fit it back into a 96 bits.

That said, can you provide the inputs that you were using? That'll help define if there is an issue or if it is expected behavior.

**Comment by ssadler** [comment]
Slightly OT, but I came across this issue in the context of blockchain also, and I'd just like to mention that floating point non determinism is only a concern if you are implementing decentralized systems, it's not a concern if you are programming smart contracts (in which case the VM should ensure determinism).

**Comment by maurolacy** [comment]
> in which case the VM should ensure determinism.

Correct. As long as your VM supports floating point, that is. CosmWasm, by the way, now [does support it](https://github.com/CosmWasm/cosmwasm/pull/1864), but didn't at the time of my original post.
